python/dgl/contrib/dis_kvstore.py

```diff
@@ -0,0 +1,487 @@
+            par_shape = shape.copy()
+            if shape[0] - server_id*count >= count:
+                par_shape[0] = count
+            else:
+                par_shape[0] = shape[0] - server_id*count
+            tensor_shape = F.tensor(par_shape)
+            msg = KVStoreMsg(
+                type=KVMsgType.INIT,
+                rank=self._client_id,
+                name=name,
+                id=tensor_shape,
+                data=threshold)
+            _send_kv_msg(self._sender, msg, server_id)
+        
+    def push(self, name, ID, data):
+        """Push message to KVServer
+
+        The push() API will partition message into different 
+        KVServer nodes automatically.
+
+        Parameters
+        ----------
+        name : str
+            data name
+        ID : tensor (mx.ndarray or torch.tensor)
+            a vector storing the global IDs
+        data : tensor (mx.ndarray or torch.tensor)
+            a matrix with the same row size of id
+        """
+        assert F.ndim(ID) == 1, 'ID must be a vector.'
+        assert F.ndim(data) == 2, 'data must be a matrix.'
+        assert F.shape(ID)[0] == F.shape(data)[0], 'The data must has the same row size with ID.'
+        group_size = [0] * self._server_count
+        numpy_id = F.asnumpy(ID)
+        for id in numpy_id:
+            # mapping data to corresponded server nodes
+            server_id = self._get_server_id(id, name)
+            group_size[server_id] += 1
+        min_idx = 0
+        max_idx = 0
+        for idx in range(self._server_count):
+            if group_size[idx] == 0:
+                continue
+            max_idx += group_size[idx]
+            range_id = ID[min_idx:max_idx]
+            range_data = data[min_idx:max_idx]
+            min_idx = max_idx
+            msg = KVStoreMsg(
+                type=KVMsgType.PUSH,
+                rank=self._client_id,
+                name=name,
+                id=range_id,
+                data=range_data)
+            _send_kv_msg(self._sender, msg, idx)
+
+    def pull(self, name, ID):
+        """Pull message from KVServer
+
+        Parameters
+        ----------
+        name : str
+            data name
+        ID : tensor (mx.ndarray or torch.tensor)
+            a vector storing the IDs
+
+        Return
+        ------
+        tensor
+            a matrix with the same row size of ID
+        """
+        assert F.ndim(ID) == 1, 'ID must be a vector.'
+        group_size = [0] * self._server_count
+        numpy_id = F.asnumpy(ID)
+        for id in numpy_id:
+            server_id = self._get_server_id(id, name)
+            group_size[server_id] += 1
+        min_idx = 0
+        max_idx = 0
+        server_count = 0
+        for idx in range(self._server_count):
+            if group_size[idx] == 0:
+                continue
+            server_count += 1
+            max_idx += group_size[idx]
+            range_id = ID[min_idx:max_idx]
```

Comment:
You assume the row Ids in `ID` is in the ascending order. Is it a right assumption?